### ISSUE TITLE
Casting not stopped with Smart Display/Speaker when using local media

### ISSUE THREAD

**Issue by greghesp** [skorokithakis/catt#236]
There are 2 symptoms to this issue, however my hunch is that they're both the same cause.  This happens on both my Google Nest Mini, as well as my Google Nest Hubs.  To test, I have the cast initiation sound switched on

### **Cast session stays connected after locally played media ends.**
**What Happens**
Casting a local file to my Nest Hub plays the file, however the cast session stays connected after the media finishes playing as per the below image.  I can tell this as the Google Nest Hub displays the cast logo as below, and for both the speaker and hub, when another file is cast to the devices, the cast initiation sound does not play.

```catt -d "Office Hub" cast C:\Users\Greg\Documents\Projects\assistant-relay\relay\bin\sounds\doorbell.mp3```

![image](https://user-images.githubusercontent.com/768908/74238886-9da48400-4cce-11ea-8e93-54f96eefb68b.png)

Attempting a stop command does not kill the session either ```catt -d "Office Hub" stop``` and leaves the Smart Display on the above screen.

Eventually, the cast session times out and the Smart Display resumes its normal operation

I have also tried this with an MP4 file, and can confirm that whilst it is playing, the stop command does not stop the cast.

**Comment by greghesp** [comment]
Interestingly, I found that whilst the display is stuck on this screen, if you use the cast tool in Chrome, it detects there is media playing on the device and lets you stop it.  This then kills the cast session running on the display or speaker.

**Comment by theychx** [comment]
@greghesp

I've implemented what we discussed. Install with:
```
pip3 install -U https://github.com/skorokithakis/catt/archive/playback-end.zip
```

Can you test this cmd?:
```
catt cast https://file-examples.com/wp-content/uploads/2018/04/file_example_WMV_480_1_2MB.wmv
```
Does that cause the same problems as you've described above?

Also, can you test whether firing multiple `catt stop` cmds at a "stalled" device makes any difference?

**Comment by greghesp** [comment]
> Also, can you test whether firing multiple `catt stop` cmds at a "stalled" device makes any difference?

That doesn't seem to do anything with the release on git

I've updated to that zip, and running `catt -d "Kitchen display" cast https://file-examples.com/wp-content/uploads/2018/04/file_example_WMV_480_1_2MB.wmv" just starts a cast session, but the file does not play

**Comment by theychx** [comment]
> but the file does not play

Yes, that was intentional. But afterwards, does the device then exibit this behaviour where `catt stop` doesn't work?

**Comment by greghesp** [comment]
> Yes, that was intentional. But afterwards, does the device then exibit this behaviour where `catt stop` doesn't work?

Ah!  No it's does not.  It starts the cast session with the cast noise and "freezes" on the black cast screen

**Comment by theychx** [comment]
Try putting the device into the "frozen" state, and then see if casting a youtube video with `catt` afterwards works.

**Comment by greghesp** [comment]
It does, although that behaviour was present before.

If I cast a YouTube video after the Default Media Player is "frozen" then it closes the Default Media Player, goes back to the display home quickly, and then fires up the YouTube app

**Comment by theychx** [comment]
> It does, although that behaviour was present before.

The new branch is not changed in any way in this regard. It just implements a blocking option for `catt cast`, and also gets rid of the lame behaviour, where you'd have to Ctrl+C `catt` yourself, after having cast a local file.

So now you can do `catt cast -b <url>` and catt will not exit until playback has ended.

**Comment by greghesp** [comment]
Ah yes, I see now:
![image](https://user-images.githubusercontent.com/768908/74382328-2281c180-4de5-11ea-93b3-2b775cbe10c8.png)

So CATT has now been killed, but the "frozen" cast session still remains

**Comment by greghesp** [comment]
I think this is an issue with the Default Media Player receiver perhaps.

I've just tried casting a Twitch stream which loads the Default Media Player, however you cannot kill the cast with CATT.

**Comment by greghesp** [comment]
@theychx Just wanted to add that I tried spamming `catt -d "device" stop` again and it did eventually work

**Comment by theychx** [comment]
Can you try running [this script](https://gist.github.com/theychx/db14d521c80147bbaa2f03801271f383) when a device has "frozen" (with device ip-address as argument)?

**Comment by greghesp** [comment]
Yes, that killed it

**Comment by theychx** [comment]
> the -b argument doesnt seem to make any difference

Thats because the shell is already blocked by `catt`s running server, so the `-b` option only has an effect when casting remote files. I should mention that in the option shorthelp, thanks.

**Comment by theychx** [comment]
> and you cannot ctrl+c to abort out of it

I'll fix that.

**Comment by greghesp** [comment]
When you see the cursor appear, I've just done `ctrl+c` and then clicked.

![ha](https://user-images.githubusercontent.com/768908/74427631-64475200-4e4f-11ea-844a-88f4fce5a2ef.gif)

**Comment by theychx** [comment]
Ok, here's the plan: I'll finish up the block option PR, and then when that is merged, I'll make a new PR that implements the workaround from my script.

Also, I need you to post the output of `catt scan`, while all of the relevent devices are on your network (you can leave out ip-addresses).

**Comment by theychx** [comment]
> When you see the cursor appear, I've just done ctrl+c and then clicked

Seems like this issue is Windows-related. And I know close to nothing about Windows.

**Comment by greghesp** [comment]
> Ok, here's the plan: I'll finish up the block option PR, and then when that is merged, I'll make a new PR that implements the workaround from my script.
> 
> Also, I need you to post the output of `catt scan`, while all of the relevant devices are on your network (you can leave out ip-addresses).

Not fussed about the IP's since they're internal

Scanning Chromecasts...
192.168.1.170 - Home Speakers - Unknown manufacturer Google Cast Group
192.168.1.185 - Kitchen display - Unknown manufacturer Google Nest Hub Max
192.168.1.9 - Living Room - Unknown manufacturer SHIELD Android TV
192.168.1.182 - Living Room Speaker - Unknown manufacturer Google Nest Mini
192.168.1.15 - Master bedroom speaker - Unknown manufacturer Google Home Mini
192.168.1.11 - Office Hub - Unknown manufacturer Google Nest Hub
192.168.1.170 - Office speaker - Unknown manufacturer Google Home Mini
192.168.1.140 - Rachs Speaker - Unknown manufacturer Google Home Mini



> > When you see the cursor appear, I've just done ctrl+c and then clicked
> 
> Seems like this issue is Windows-related. And I know close to nothing about Windows.

Let me fire up `cmder` and see if that works

**Comment by greghesp** [comment]
It does not.  Hmm, how bizzare.

**Comment by theychx** [comment]
Does the "SHIELD Android TV" device also exibit the freezing behaviour?

**Comment by greghesp** [comment]
I haven't tried that device.  I'll check it out in a bit and report back

**Comment by greghesp** [comment]
The Shield is actually telling me the file isn't compatible even though its an mp4 file... bizarre.
It loads the default media receiver up though, however I can close that fine with `catt stop`

**Comment by theychx** [comment]
I've implemented the fix for the stop cmd issues. Please test.
```
pip3 install -U https://github.com/skorokithakis/catt/archive/stop-fix.zip
```

**Comment by greghesp** [comment]
I don't think this is now an issue @theychx 

When casting a local file, once it is closed, it now exits automatically, however it is now firing up a blank receiver and then immediately closing it  (This was the fix put in place, but seems the latest Nest Hub update resolved this itself)
